use Nerdbank.GitVersioning for versioning

ctaggart opened the pull request:
This updates the build to use https://github.com/AArnott/Nerdbank.GitVersioning so that it is easy to fork, test, and contribute to this project. Builds not on master will be prerelease with the git short commit as the prerelease part. Please ask questions. I'll probably have to explain more in a blog post.

granicz commented:
Yes, some extra detail would help here to review. Also, instead of removing the AssemblyInfo entries, we should find a way to keep them.

ctaggart commented:
I'll start with the assembly information. For your `0.4.144` release it had this:
![image](https://user-images.githubusercontent.com/80104/55972560-9210a880-5c38-11e9-861b-cf5e1e148b8a.png)

The last build in this pull request produced:
![image](https://user-images.githubusercontent.com/80104/55972735-ec116e00-5c38-11e9-87ae-e3199ddad453.png)

I think the only thing missing from the assembly info at the moment is `[assembly: AssemblyCompany("IntelliFactory")]`. Fake was doing an `AssemblyInfo.Company "IntelliFactory"` for that. It is probably just `<Company>IntelliFactory</Company>` in the project files. I'll have a look. 

With this pull request, the new .NET project system and Nerdbank.GitVersioning package references are creating the assembly info. I added the Paket equivalent [to this](https://github.com/ctaggart/froto/blob/81263c5b8ca897ce1169072495f6a9849f0b2991/Parser/Froto.Parser.fsproj#L24):
``` xml
<PackageReference Include="Nerdbank.GitVersioning" Version="2.3.138" PrivateAssets="All" />
``` 
which in Paket is:
```
nuget Nerdbank.GitVersioning copy_local: true
```

ctaggart commented:
I added the company info back and now the assembly info for the last commit is this:

![image](https://user-images.githubusercontent.com/80104/55976631-b9b83e80-5c41-11e9-894a-1d80b056115c.png)

One of the cool things is that you are going to get the same version if you build that commit locally as well. Just try `dotnet build` or build from Visual Studio and have a look. The build numbers are deterministic. If I added another CI server, you would get the same build number on it as well for the same commit.

ctaggart commented:
@granicz, I wrote up that blog post:
http://blog.ctaggart.com/2019/04/automate-your-net-versioning-with.html

Tarmil commented:
Thanks for the writeup! I agree that this looks like a better versioning system than what we currently have.

I guess this would have to be merged as 0.5 rather than 0.4, otherwise it would create lower version numbers than what's already published.

ctaggart commented:
Sounds good. I bumped it up to 0.5. I recommend squash merging.

granicz commented:
Cool, thanks @ctaggart!
